File: templates/Features/SuspendAndResumeVB/Services/SuspendAndResumeService.vb
Imports System
Imports System.Threading.Tasks

Imports Windows.ApplicationModel.Activation
Imports Windows.Storage
Imports Windows.UI.Xaml.Controls

Imports Param_RootNamespace.Activation
Imports Param_RootNamespace.Helpers

Namespace Services
    Friend Class SuspendAndResumeService
        Inherits ActivationHandler(Of LaunchActivatedEventArgs)

        ' TODO WTS: For more information regarding the application lifecycle and how to handle suspend and resume, please see:
        ' Documentation: https://docs.microsoft.com/windows/uwp/launch-resume/app-lifecycle

        Private Const StateFilename As String = "suspensionState"

        ' TODO WTS: This event is fired just before the app enters in background. Subscribe to this event if you want to save your current state.
        Public Event OnBackgroundEntering As EventHandler(Of OnBackgroundEnteringEventArgs)

        Public Async Function SaveStateAsync() As Task
            Dim suspensionState = New SuspensionState() With {
                .SuspensionDate = DateTime.Now
            }

            Dim target As Type = Nothing
            
            if OnBackgroundEnteringEvent IsNot Nothing Then
                target = OnBackgroundEnteringEvent.Target.GetType()
            EndIf

            Dim onBackgroundEnteringArgs = New OnBackgroundEnteringEventArgs(suspensionState, target)

            RaiseEvent OnBackgroundEntering(Me, onBackgroundEnteringArgs)

            Await ApplicationData.Current.LocalFolder.SaveAsync(StateFilename, onBackgroundEnteringArgs)
        End Function

        Protected Overrides Async Function HandleInternalAsync(args As LaunchActivatedEventArgs) As Task
            Await RestoreStateAsync()
        End Function

        Protected Overrides Function CanHandleInternal(args As LaunchActivatedEventArgs) As Boolean
            Return args.PreviousExecutionState = ApplicationExecutionState.Terminated
        End Function

        Private Async Function RestoreStateAsync() As Task
            Dim saveState = Await ApplicationData.Current.LocalFolder.ReadAsync(Of OnBackgroundEnteringEventArgs)(StateFilename)
        End Function
    End Class
End Namespace

File: templates/Features/SuspendAndResume/Services/SuspendAndResumeService.cs
using System;
using System.Threading.Tasks;

using Windows.ApplicationModel.Activation;
using Windows.Storage;
using Windows.UI.Xaml.Controls;

using Param_RootNamespace.Activation;
using Param_RootNamespace.Helpers;

namespace Param_ItemNamespace.Services
{
    internal class SuspendAndResumeService : ActivationHandler<LaunchActivatedEventArgs>
    {
        //// TODO WTS: For more information regarding the application lifecycle and how to handle suspend and resume, please see:
        //// Documentation: https://docs.microsoft.com/windows/uwp/launch-resume/app-lifecycle

        private const string StateFilename = "SuspendAndResumeState";

        // TODO WTS: This event is fired just before the app enters in background. Subscribe to this event if you want to save your current state.
        public event EventHandler<OnBackgroundEnteringEventArgs> OnBackgroundEntering;

        public async Task SaveStateAsync()
        {
            var suspensionState = new SuspensionState()
            {
                SuspensionDate = DateTime.Now
            };

            var target = OnBackgroundEntering?.Target.GetType();
            var onBackgroundEnteringArgs = new OnBackgroundEnteringEventArgs(suspensionState, target);

            OnBackgroundEntering?.Invoke(this, onBackgroundEnteringArgs);

            await ApplicationData.Current.LocalFolder.SaveAsync(StateFilename, onBackgroundEnteringArgs);
        }

        protected override async Task HandleInternalAsync(LaunchActivatedEventArgs args)
        {
            await RestoreStateAsync();
        }

        protected override bool CanHandleInternal(LaunchActivatedEventArgs args)
        {
            return args.PreviousExecutionState == ApplicationExecutionState.Terminated;
        }

        private async Task RestoreStateAsync()
        {
            var saveState = await ApplicationData.Current.LocalFolder.ReadAsync<OnBackgroundEnteringEventArgs>(StateFilename);
        }
    }
}
